content/ch-quantum-hardware/ac-Stark-shift.ipynb

```diff
@@ -0,0 +1,465 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from scipy.optimize import leastsq,minimize, curve_fit\n",
+    "\n",
+    "# samples need to be multiples of 16 to accommodate the hardware limitations\n",
+    "def get_closest_multiple_of_16(num):\n",
+    "    return int(num + 8 ) - (int(num + 8 ) % 16)\n",
+    "\n",
+    "# lorentzian function\n",
+    "def lorentzian(f, f0, k, a, offs):\n",
+    "    return a*k/(2*np.pi)/((k/2)**2+(f-f0)**2)+offs\n",
+    "\n",
+    "# fit a lorentizan function\n",
+    "def fit_lorentzian(freqs,values):\n",
+    "    p0=[freqs[np.argmax(values)],(freqs[-1]-freqs[0])/2,max(values),0]\n",
+    "    bounds=([freqs[0],0,0,-np.inf],[freqs[-1],freqs[-1]-freqs[0],np.inf,np.inf])\n",
+    "    popt,pcov=curve_fit(lorentzian, freqs, values, p0=p0, bounds=bounds)\n",
+    "    return popt,pcov\n",
+    "\n",
+    "# Gaussian function\n",
+    "def gaussian(f, f0, sigma, a, offs):\n",
+    "    return a*np.exp(-(f-f0)**2/(2*sigma**2))+offs\n",
+    "\n",
+    "# fit a Gaussian function\n",
+    "def fit_gaussian(freqs,values):\n",
+    "    p0=[freqs[np.argmax(values)],(freqs[-1]-freqs[0])/2,max(values),0]\n",
+    "    bounds=([freqs[0],0,0,-np.inf],[freqs[-1],freqs[-1]-freqs[0],np.inf,np.inf])\n",
+    "    popt,pcov=curve_fit(gaussian, freqs, values, p0=p0, bounds=bounds)\n",
+    "    return popt,pcov\n",
+    "\n",
+    "# normalize the data points to fall in the range of [0,1]\n",
+    "def normalize(a):\n",
+    "    a= a-min(a)\n",
+    "    return a/max(a)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1.  ac-Stark Shifting the qubit\n",
+    "In order to ac-Stark shift the qubit we need to populate the resonator with photons using an on-resonance drive. For a drive amplitude $\\epsilon$, and a resonator decay rate of $\\kappa$, the number of photons in the resonator $\\bar{n}=\\langle a^\\dagger a \\rangle = \\frac{\\epsilon^2}{\\Delta^2 +(\\kappa/2)^2}$. As a reminder $\\tilde{\\omega}_q=\\omega_q + \\frac{g^2}{\\Delta} + \\delta \\omega_q$ where the shift in frequency due to ac_Stark shift is $\\delta \\omega_q = \\frac{2g^2}{\\Delta} \\bar{n}$. Since $\\Delta=\\omega_q-\\omega_r<0$ the qubit frequency gets smaller as we increase the of photons in the resonator"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qiskit import pulse            # This is where we access all of our Pulse features!\n",
+    "from qiskit.pulse import Play, Acquire\n",
+    "from qiskit.pulse import pulse_lib\n",
+    "import numpy as np\n",
+    "\n",
+    "qubit=0   # qubit used in our experiment\n",
+    "\n",
+    "inst_sched_map = backend_defaults.instruction_schedule_map\n",
+    "# Get the default measurement pulse sequence\n",
+    "measure = inst_sched_map.get('measure', qubits=backend_config.meas_map[0])  \n",
+    "\n",
+    "qubit_drive_sigma = 100e-9\n",
```

Comment:
Can you explain where these values come from please?

Reply:
The choice was a bit arbitrary, but I'll write a comment

Reply:
Thank you